internal/controller/nonadminrestore_controller.go

```diff
@@ -0,0 +1,339 @@
+
+// Reconcile is part of the main kubernetes reconciliation loop which aims to
+// move the current state of the cluster closer to the desired state,
+// defined in NonAdminRestore object Spec.
+func (r *NonAdminRestoreReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	logger := log.FromContext(ctx)
+	logger.V(1).Info("NonAdminRestore Reconcile start")
+
+	nar := &nacv1alpha1.NonAdminRestore{}
+	err := r.Get(ctx, req.NamespacedName, nar)
+	if err != nil {
+		if apierrors.IsNotFound(err) {
+			logger.V(1).Info(err.Error())
+			return ctrl.Result{}, nil
+		}
+		logger.Error(err, "Unable to fetch NonAdminRestore")
+		return ctrl.Result{}, err
+	}
+
+	if !nar.DeletionTimestamp.IsZero() {
+		updatedPhase := updateNonAdminPhase(&nar.Status.Phase, nacv1alpha1.NonAdminPhaseDeleting)
+		updatedCondition := meta.SetStatusCondition(&nar.Status.Conditions,
+			metav1.Condition{
+				Type:    string(constant.NonAdminConditionDeleting),
+				Status:  metav1.ConditionTrue,
+				Reason:  "DeletionPending",
+				Message: "restore accepted for deletion",
+			},
+		)
+		if updatedPhase || updatedCondition {
+			if err := r.Status().Update(ctx, nar); err != nil {
+				logger.Error(err, statusUpdateError)
+				return ctrl.Result{}, err
+			}
+			logger.V(1).Info("NonAdminRestore status marked for deletion")
+		} else {
+			logger.V(1).Info("NonAdminRestore status unchanged during deletion")
+		}
+
+		veleroRestore, err := function.GetVeleroRestoreByLabel(ctx, r.Client, r.OADPNamespace, nar.Status.UUID)
+		if err != nil {
+			if !apierrors.IsNotFound(err) {
+				logger.Error(err, "Unable to fetch Velero Restore")
+				return ctrl.Result{}, err
+			}
+		} else {
+			// 	TODO when/why Velero Restore has finalizer?
+			// 	TODO how to safely remove Velero Restore with finalizer?
+			for _, finalizer := range veleroRestore.GetFinalizers() {
+				controllerutil.RemoveFinalizer(veleroRestore, finalizer)
+			}
+			// TODO does this change generation? need to refetch?
+			if err := r.Update(ctx, veleroRestore); err != nil {
+				return ctrl.Result{}, err
+			}
+			err = r.Delete(ctx, veleroRestore)
+			if err != nil {
+				logger.Error(err, "Unable to delete Velero Restore")
+				return ctrl.Result{}, err
+			}
+		}
+
+		controllerutil.RemoveFinalizer(nar, constant.NonAdminRestoreFinalizerName)
+		// TODO does this change generation? need to refetch?
+		if err := r.Update(ctx, nar); err != nil {
+			return ctrl.Result{}, err
+		}
+		logger.V(1).Info("NonAdminRestore Reconcile exit")
+		return ctrl.Result{}, nil
+	}
+
+	reconcileSteps := []nonAdminRestoreReconcileStepFunction{
+		r.init,
+		r.validateSpec,
+		r.setUUID,
+		r.setFinalizer,
+		r.createVeleroRestore,
+	}
+	for _, step := range reconcileSteps {
+		requeue, err := step(ctx, logger, nar)
+		if err != nil {
+			return ctrl.Result{}, err
+		} else if requeue {
+			// TODO needed?
```

Comment:
If none of the reconcile steps have requeue logic then no, then also step can be simpler without requeue bool ?